Repleace Poco::Mutex, Poco::FastMutex with elements from header <mutex>

picatess opened the pull request:
Fixes #15288 

__Context__

The aim of this issue is to move `Poco` mutexes to `std` mutexes. The way the two systems map is described in the issue itself.  Note that anything associated with `Poco/RWLock` was not touched.


### For tester

* Code review carefully. I did leave the original structure in place. In some cases I changed the manual locking to either use a `lock_guard` or a `unique_lock`. Assure yourself that it makes sense in these situations. Note that we should in the future probably replace the manual, direct locking with a `unique_lock` in the future, but this should really be in a separate ticket. 

* This is a quite broad change, so one thing that could be done is to run the script below and confirm that there is no error. `WorkspacesGroup` was one of the few tricky things since it has a `recursive_mutex` and essentially is self-deleting. This did cause a resource leak which had been undetected by `Poco` (and `std` in Release).
```python
# Create a sample group workspace
ws1 = CreateSampleWorkspace()
ws2 = CreateSampleWorkspace()
ws = GroupWorkspaces(InputWorkspaces=[ws1,ws2])


# Remove the child workspaces from the ADS
mtd.remove("ws1")
mtd.remove("ws2")
```

mantid-builder commented:
Thanks for the changes. Here's a few things that we'll use to start the review process:

#### Main Reviewer ####

Please comment on the following:

##### Pull Request #####

- [ ] Does the title summarize the changes, i.e. it's not the automatically generated one?
- [ ] Is an issue link included/required, e.g. "Fixes #XXXX"? If there is an issue link do the changes look sensible and consistent with issue description?
- [ ] Are there adequate testing instructions?
- [ ] A link to the updated release notes has been provided or no release note updates were necessary?

##### Code Review #####

- [ ] Is the code of an acceptable quality?
  - [ ] Does the code conform to the coding standards? Is it well structured with small focussed classes/methods/functions?
  - [ ] Are there unit/system tests in place? Are the unit tests small and test the class(es) in isolation from one another?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
  - [ ] How do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant documentation been added/updated?
  - [ ] Is user-facing documentation written in a user-friendly manner?
  - [ ] Has developer documentation been updated if required?

***

- Does everything look good? Comment with the ship it emoji but don't merge.


#### Second Reviewer (`@gatekeepers`) ####

- [ ] Have adequate tests, both success & failure cases, been performed?,
- [ ] Has the code been reviewed?,
- [ ] Ready to go? Hit `Merge`!

quantumsteve commented:
Would it be better to use the standard library directly instead of through typedefs in MultiThreaded.h?

picatess commented:
The issue that we see on OSX can be reproduced on Windows in debug mode. It appears a `Workspacegroup` ( which contains a recursive mutex) is being deleted with a locked mutex. This causes the assertion.

The mutex was locked with a lockguard, so in terms of precaution it is handled ok. The issue is rather that the workspace is responsible for its  own destruction, or at least the `Workspacegroup::workspaceDeleteHandle` is still operating when the workspace group is being deleted, hence the scope of the lock guard still exists and the mutex flag is still locked when the workspace is being destroyed. 

![call_stack_recursive_mutex_issue](https://cloud.githubusercontent.com/assets/8955784/13224292/bbb4ada6-d97f-11e5-9fe8-bb45a267fe6d.png)

quantumsteve commented on Framework/API/inc/MantidAPI/Projection.h:
```diff
@@ -8,7 +8,7 @@
 #include <boost/shared_ptr.hpp>
 
-using namespace Mantid::Kernel;
+// using namespace Mantid::Kernel;
```
Thank you for catching and fixing this! :+1: 

Would you please delete the commented out line.

quantumsteve commented on Framework/API/inc/MantidAPI/BoxController.h:
```diff
@@ -286,7 +285,7 @@ class DLLExport BoxController {
    */
   void trackNumBoxes(size_t depth) {
-    m_mutexNumMDBoxes.lock();
+    std::lock_guard<std::mutex> lock(m_mutexNumMDBoxes);
```
RAII is superior to locking and unlocking the mutex :+1:

quantumsteve commented on Framework/DataObjects/src/BoxControllerNeXusIO.cpp:
```diff
@@ -466,7 +466,7 @@ void BoxControllerNeXusIO::closeFile() {
     // lock file
-    Poco::ScopedLock<Poco::FastMutex> _lock(m_fileMutex);
+    std::lock_guard<std::mutex> _lock(m_fileMutex);
     {
```
What purpose do these parenthesis serve? should they be removed?

quantumsteve commented on Framework/Kernel/src/DiskBuffer.cpp:
```diff
@@ -52,19 +52,18 @@ void DiskBuffer::toWrite(ISaveable *item) {
   {
     // forget old memory size
-    m_mutex.lock();
+    std::unique_lock<std::mutex> uniqueLock(m_mutex);
```
Would you be better off with a `std::lock_guard` that goes out of scope one line later at the `}`?

picatess replied:
The `std::lock_guard` cannot be unlocked manually, the `std::unique_lock` can. In the original code mutex needs to be unlocked before `item->setBufferSize(newMemorySize);` line.

quantumsteve commented:
Aside from the commented-out line and (previously existing) extra parentheses, the changes look good. I double-checked that `Poco::Mutex` changed to `std::recursive_mutex`. and `Poco::FastMutex` changed to `std::mutex`. Unlocking the GroupWorkspace looks reasonable and fixes the assertion failure on OS X. After those two things and the merge conflicts, and I believe it will be ready to merge into master.

I commented on one place where a `lock_guard` might make more sense than a `unique_lock`, but am not familiar enough with the that function to be confident.

picatess commented:
@quantumsteve 
Thanks for checking the PR. Following your comments I have
* removed the extra parenthesis
* removed the uncommented line

As mentioned above, the `unique_lock` is required since  we actively want to unlock the mutex later on.

I cannot think of  a strong reason to perform manual locking instead of applying `std::unique_lock` to the bare mutex operations that still exist in the code. I will create another maintenance ticket for this, but I did not want to do this with this PR.

quantumsteve commented:
Last changeset looks good. Limiting the scope makes sense because this PR is already quite large.
:shipit:

FedeMPouzols commented:
Good changes and carefully reviewed.
